File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
  const map = [[],[],
               ['a','b','c'],
               ['d','e','f'],
               ['g','h','i'],
               ['j','k','l'],
               ['m','n','o'],
               ['p','q','r','s'],
               ['t','u','v'],
               ['w','x','y','z']]
  const combine = aoa => {
    if (aoa.length === 0) return []
    if (aoa.length === 1) return aoa[0]
    const [h, ...t] = aoa
    return [].concat(...(combine(t).map(s => [...(h.map(c => c+s))])))
  }
  return combine(digits.split('').map(ch => map[+ch]))
};

File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.py
class Solution:
  def letterCombinations(self, digits: str) -> List[str]:
    dic = {
      "2": "abc",
      "3": "def",
      "4": "ghi",
      "5": "jkl",
      "6": "mno",
      "7": "pqrs",
      "8": "tuv",
      "9": "wxyz"
    }
    def combine(aoa: List[Iterable[str]]) -> List[str]:
      if not aoa: return []
      if len(aoa) == 1: return aoa[0]
      head, *tail = aoa
      return [h+t for h in head for t in combine(tail)]
    return combine(list(map(lambda d: dic[d], digits)))